STYLE_GUIDE.md

```diff
@@ -0,0 +1,461 @@
+  )
+
+// Not OK
+type Signed is ( I8
+              | I16
+              | I32
+              | I64
+              | I128 )
+
+// Not OK
+type Signed is
+  (I8
+  |I16
+  |I32
+  |I64
+  |I128)
+```
+
+### Array Literals
+
+Multiline arrays generally have a newline between each entry, though logical exceptions exist (such as tables).
+
+```pony
+// OK
+let ns = [as USize: 1; 2; 3]
+
+// OK
+let strs = ["a"; "b"; "c"]
+
+// OK
+let ns =
+  [ as USize:
+    1
+    2
+    3 ]
+
+// OK
+let ns =
+  [ as USize:
+    1
+    2
+    3
+  ]
+
+// Not OK
+let ns =
+  [as USize:
+   1
+   2
+   3
+  ]
+
+// Not OK
+let ns = [ as USize:
+    1
+    2
+    3
+]
+
+// Not OK
+let ns = [ as USize:
+          1
+          2
+          3 ]
+```
+
+### Control Structures
+
+- Control structures are aligned so that their keywords (if, else, for, try, end, etc.) are at the same indentation level.
+
+  ```pony
+  if cond1 then e1
+  elif cond2 then e2
+  else e3
+  end
+
+  if cond1 then
+    e1
+  elif cond2 then
+    e2
+  else
+    e3
+  end
+
+  for name in ["Bob"; "Fred"; "Sarah"].values() do
+    env.out.print(name)
+  end
+  ```
+
+- Small expressions may be placed on a single line.
+
+  ```pony
+  if cond then e1 else e2 end
+  ```
+
+- Match expression cases are __not__ indented.
+
+  ```pony
+  match (x, y)
+  | (None, _) => "none"
+  | (let s: String, 2) => s + " two"
+  | (let s: String, 3) => s + " three"
+  | (let s: String, let u: U32) if u > 14 => s + " other big integer"
+  | (let s: String, _) => s + " other small integer"
+  else "something else"
+  end
+
+  match alive
+  | true =>
+    Debug.err("Beep boop!")
+    this.forever()
+  | false =>
+    Debug.err("Ugah!")
+    None
+  end
+  ```
+
+- Match expressions __may not__ be on a single line.
+
+  ```pony
+  // OK
+  match x
+  | let n: U8 => n.string()
+  else "0"
+  end
+
+  // Not OK
+  match x | let n: U8 => n.string() else "0" end
+  ```
+
+### Assignment
+
+Assignment of a multiline expression is indented below the `=` symbol.
+
+```pony
+// OK
+var x =
+  if friendly then
+    "Hello"
+  else
+    false
+  end
+
+// OK
+let output =
+  recover String(
+    file_name.size()
+      + file_linenum.size()
+      + file_linepos.size()
+      + msg.size()
+      + 4)
+  end
+
+// Not OK
+var x = if friendly then
+  "Hello"
+else
+  false
+end
+```
+
+### Function Declarations
+
+- Partial functions are denoted with a `?` separated from other symbols by one space.
+
+  ```pony
+  // OK
+  fun foo(): Foo^ ? =>
+  // ...
+
+  // OK
+  fun foo() ? =>
+  // ...
+
+  // Not OK
+  fun foo(): Foo^? =>
+  // ...
+
+  // Not OK
+  fun foo()? =>
+  // ...
+  ```
+
+- Multiline function parameters are each listed on a separate line. The return type preceded by a `:` is placed at the same indentation level followed by a question mark (if the function is partial). The `=>` is placed on a separate line between the parameter list and the function body at the same indentation level as the `fun` keyword.
+
```

Comment:
What about a case where the parameters all fit on one line, but then the return type has to be wrapped? Would something like this be OK?
```
  fun tag _init(typ': ValueType, default': (Value | None))
    : (ValueType, Value, Bool) ?
  =>
    // ...
```